### ISSUE TITLE
Note about unusual stride situations in dev docs

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#6165]
On its face, strides are pretty simple: given strides `s`, the memory location of the indices `x` is `sum(s[i] * x[i] for i in dim)`. It's also fairly simple to say when strides are contiguous: that is when `s[i] == product(s[j] for j in range(0, i))`. And then the standard application of strides is to let you get non-contiguous views into an underlying contiguous tensor.

But there are also some weirder, less obvious stride situations:

Broadcasting is achieved by having zero strides. So, e.g., 

```
>>> torch.tensor(range(0,3)).expand(2,3)

 0  1  2
 0  1  2
[torch.LongTensor of size (2,3)]
>>> torch.tensor(range(0,3)).as_strided(size=(2,3), stride=(0,1))

 0  1  2
 0  1  2
[torch.LongTensor of size (2,3)]
```

And the technical definition of strides permits weird, overlapping striding, e.g.,

```
>>> torch.tensor(range(0,7)).as_strided(size=(2,2,2), stride=(3,2,1))

(0 ,.,.) = 
  0  1
  2  3

(1 ,.,.) = 
  3  4
  5  6
[torch.LongTensor of size (2,2,2)]
```

It would be good to have a developer doc describing all of this, as well as adding functions making it easier to test for certain stride situations, to make it easier to work with external libraries. An example is in the cuFFT interface https://github.com/pytorch/pytorch/pull/6118 where only non-overlapping striding is supported.

CC @SsnL
